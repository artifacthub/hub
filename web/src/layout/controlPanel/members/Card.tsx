import classnames from 'classnames';
import isUndefined from 'lodash/isUndefined';
import React, { useContext, useRef, useState } from 'react';
import { BsThreeDotsVertical } from 'react-icons/bs';
import { FaSignOutAlt, FaUser, FaUserMinus } from 'react-icons/fa';
import { IoMdCloseCircle } from 'react-icons/io';

import { API } from '../../../api';
import { AppCtx, unselectOrg } from '../../../context/AppCtx';
import useOutsideClick from '../../../hooks/useOutsideClick';
import { AuthorizerAction, ErrorKind, Member } from '../../../types';
import alertDispatcher from '../../../utils/alertDispatcher';
import Modal from '../../common/Modal';
import ActionBtn from '../ActionBtn';
import styles from './Card.module.css';

interface Props {
  member: Member;
  membersNumber: number;
  onSuccess: () => void;
  onAuthError: () => void;
}

const MemberCard = (props: Props) => {
  const { ctx, dispatch } = useContext(AppCtx);
  const [isDeletingMember, setIsDeletingMember] = useState(false);
  const dropdownMenu = useRef(null);
  const [dropdownMenuStatus, setDropdownMenuStatus] = useState<boolean>(false);
  const [modalStatus, setModalStatus] = useState<boolean>(false);

  const closeDropdown = () => {
    setDropdownMenuStatus(false);
  };

  useOutsideClick([dropdownMenu], dropdownMenuStatus, closeDropdown);

  async function deleteMember() {
    try {
      setIsDeletingMember(true);
      await API.deleteOrganizationMember(ctx.prefs.controlPanel.selectedOrg!, props.member.alias);
      setIsDeletingMember(false);
      if (props.member.alias === ctx.user!.alias) {
        dispatch(unselectOrg());
      } else {
        props.onSuccess();
      }
    } catch (err) {
      setIsDeletingMember(false);
      if (err.kind !== ErrorKind.Unauthorized) {
        let errorMessage = 'An error occurred removing member from the organization, please try again later.';
        if (err.kind === ErrorKind.Forbidden) {
          errorMessage = 'You do not have permissions to remove members from the organization.';
        }
        alertDispatcher.postAlert({
          type: 'danger',
          message: errorMessage,
        });
      } else {
        props.onAuthError();
      }
    }
  }

  const isUser = props.member.alias === ctx.user!.alias;

  const getFullName = (): string => {
    let fullName = '';
    if (props.member.firstName) {
      fullName += `${props.member.firstName} `;
    }
    if (props.member.lastName) {
      fullName += props.member.lastName;
    }
    return fullName;
  };

  return (
    <div className="col-12 col-xxl-6 py-sm-3 py-2" data-testid="memberCard">
      <div className="card h-100">
        <div className="card-body d-flex flex-column h-100">
          <div className="d-flex flex-row w-100 justify-content-between align-items-start">
            <div
              className={`d-flex align-items-center justify-content-center p-1 overflow-hidden mr-2 ${styles.imageWrapper} imageWrapper`}
            >
              <FaUser className={styles.image} />
            </div>

            <div className="flex-grow-1">
              <div className="d-flex flex-row align-items-start">
                <div className="h5 mb-1">
                  {props.member.firstName || props.member.lastName ? getFullName() : props.member.alias}
                </div>
                {!isUndefined(props.member.confirmed) && !props.member.confirmed && (
                  <div className={classnames('ml-3', { 'mr-3': props.membersNumber > 1 })}>
                    <span className="badge badge-warning">Invitation not accepted yet</span>
                  </div>
                )}
              </div>
              <div className="h6 text-muted mr-1 font-italic">{props.member.alias}</div>
            </div>

            {props.membersNumber > 1 && (
              <>
                {modalStatus && (
                  <Modal
                    className={`d-inline-block ${styles.modal}`}
                    closeButton={
                      <>
                        <button
                          className={`btn btn-sm btn-light text-uppercase ${styles.btnLight}`}
                          onClick={() => setModalStatus(false)}
                        >
                          <div className="d-flex flex-row align-items-center">
                            <IoMdCloseCircle className="mr-2" />
                            <span>Cancel</span>
                          </div>
                        </button>

                        <button
                          data-testid="leaveOrRemoveBtn"
                          className="btn btn-sm btn-danger ml-3"
                          onClick={(e) => {
                            e.preventDefault();
                            deleteMember();
                          }}
                          disabled={isDeletingMember}
                        >
                          <div className="d-flex flex-row align-items-center text-uppercase">
                            {isDeletingMember ? (
                              <>
                                <span className="spinner-grow spinner-grow-sm" role="status" aria-hidden="true" />
                                <span className="ml-2">{isUser ? 'Leaving...' : 'Removing...'}</span>
                              </>
                            ) : (
                              <>
                                {isUser ? (
                                  <FaSignOutAlt className={`mr-2 ${styles.btnIcon}`} />
                                ) : (
                                  <FaUserMinus className={`mr-2 ${styles.btnIcon}`} />
                                )}
                                <span>{isUser ? 'Leave' : 'Remove'}</span>
                              </>
                            )}
                          </div>
                        </button>
                      </>
                    }
                    header={
                      <div className={`h3 flex-grow-1 m-2 ${styles.title}`}>
                        {isUser ? 'Leave ' : 'Remove from '} organization
                      </div>
                    }
                    onClose={() => setModalStatus(false)}
                    open
                  >
                    <div className="mt-3 mw-100 text-center">
                      <p>
                        {isUser
                          ? 'Are you sure you want to leave this organization?'
                          : 'Are you sure you want to remove this member from this organization?'}
                      </p>
                    </div>
                  </Modal>
                )}

                <div className="ml-auto">
                  <div
                    ref={dropdownMenu}
                    className={classnames('dropdown-menu dropdown-menu-right p-0', styles.dropdownMenu, {
                      show: dropdownMenuStatus,
                    })}
                  >
                    <div className={`arrow ${styles.arrow}`} />

                    {isUser ? (
                      <button
                        data-testid="leaveOrRemoveModalBtn"
                        className="dropdown-item btn btn-sm rounded-0 text-secondary"
                        onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                          e.preventDefault();
                          closeDropdown();
                          setModalStatus(true);
                        }}
                      >
                        <div className="d-flex flex-row align-items-center">
                          <FaSignOutAlt className={`mr-2 ${styles.btnIcon}`} />
                          <span>Leave</span>
                        </div>
                      </button>
                    ) : (
                      <ActionBtn
                        testId="leaveOrRemoveModalBtn"
                        className="dropdown-item btn btn-sm rounded-0 text-secondary"
                        onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                          e.preventDefault();
                          closeDropdown();
                          setModalStatus(true);
                        }}
                        action={AuthorizerAction.DeleteOrganizationMember}
                      >
                        <>
                          <FaUserMinus className={`mr-2 ${styles.btnIcon}`} />
                          <span>Remove</span>
                        </>
                      </ActionBtn>
                    )}
                  </div>

                  <button
                    className={`btn btn-light p-0 text-secondary text-center ${styles.btnDropdown}`}
                    onClick={() => setDropdownMenuStatus(true)}
                  >
                    <BsThreeDotsVertical />
                  </button>
                </div>
              </>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default MemberCard;
