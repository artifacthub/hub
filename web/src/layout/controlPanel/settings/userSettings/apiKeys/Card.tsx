import classnames from 'classnames';
import moment from 'moment';
import React, { useRef, useState } from 'react';
import { BsThreeDotsVertical } from 'react-icons/bs';
import { FaPencilAlt, FaTrashAlt } from 'react-icons/fa';
import { IoMdCloseCircle } from 'react-icons/io';

import { API } from '../../../../../api';
import useOutsideClick from '../../../../../hooks/useOutsideClick';
import { APIKey, ErrorKind } from '../../../../../types';
import alertDispatcher from '../../../../../utils/alertDispatcher';
import ButtonCopyToClipboard from '../../../../common/ButtonCopyToClipboard';
import Modal from '../../../../common/Modal';
import styles from './Card.module.css';

interface ModalStatus {
  open: boolean;
  apiKey?: APIKey;
}

interface Props {
  apiKey: APIKey;
  setModalStatus: React.Dispatch<React.SetStateAction<ModalStatus>>;
  onSuccess: () => void;
  onAuthError: () => void;
}

const APIKeyCard = (props: Props) => {
  const [isDeleting, setIsDeleting] = useState(false);
  const [dropdownMenuStatus, setDropdownMenuStatus] = useState<boolean>(false);
  const dropdownMenu = useRef(null);
  const [deletionModalStatus, setDeletionModalStatus] = useState<boolean>(false);

  const closeDropdown = () => {
    setDropdownMenuStatus(false);
  };

  useOutsideClick([dropdownMenu], dropdownMenuStatus, closeDropdown);

  async function deleteAPIKey() {
    try {
      setIsDeleting(true);
      await API.deleteAPIKey(props.apiKey.apiKeyId!);
      setIsDeleting(false);
      props.onSuccess();
    } catch (err) {
      setIsDeleting(false);
      if (err.kind === ErrorKind.Unauthorized) {
        props.onAuthError();
      } else {
        alertDispatcher.postAlert({
          type: 'danger',
          message: 'An error occurred deleting the API key, please try again later.',
        });
      }
    }
  }

  return (
    <div className="col-12 col-xxl-6 py-sm-3 py-2" data-testid="APIKeyCard">
      <div className="card h-100">
        <div className="card-body d-flex flex-column h-100">
          <div className="d-flex flex-row w-100 justify-content-between">
            <div className={`h5 mb-1 mr-2 ${styles.titleCard}`}>{props.apiKey.name}</div>
            {deletionModalStatus && (
              <Modal
                className={`d-inline-block ${styles.modal}`}
                closeButton={
                  <>
                    <button
                      className={`btn btn-sm btn-light text-uppercase ${styles.btnLight}`}
                      onClick={() => setDeletionModalStatus(false)}
                    >
                      <div className="d-flex flex-row align-items-center">
                        <IoMdCloseCircle className="mr-2" />
                        <span>Cancel</span>
                      </div>
                    </button>

                    <button
                      data-testid="deleteAPIKeyBtn"
                      className="btn btn-sm btn-danger ml-3"
                      onClick={(e) => {
                        e.preventDefault();
                        closeDropdown();
                        deleteAPIKey();
                      }}
                      disabled={isDeleting}
                    >
                      <div className="d-flex flex-row align-items-center text-uppercase">
                        {isDeleting ? (
                          <>
                            <span className="spinner-grow spinner-grow-sm" role="status" aria-hidden="true" />
                            <span className="ml-2">Deleting...</span>
                          </>
                        ) : (
                          <>
                            <FaTrashAlt className={`mr-2 ${styles.btnDeleteIcon}`} />
                            <span>Delete</span>
                          </>
                        )}
                      </div>
                    </button>
                  </>
                }
                header={<div className={`h3 m-2 flex-grow-1 ${styles.title}`}>Delete API key</div>}
                onClose={() => setDeletionModalStatus(false)}
                open
              >
                <div className="mt-3 mw-100 text-center">
                  <p>Are you sure you want to remove this API key?</p>
                </div>
              </Modal>
            )}

            <div className="ml-auto">
              <div
                ref={dropdownMenu}
                className={classnames('dropdown-menu dropdown-menu-right p-0', styles.dropdownMenu, {
                  show: dropdownMenuStatus,
                })}
              >
                <div className={`arrow ${styles.arrow}`} />

                <button
                  data-testid="updateAPIKeyBtn"
                  className="dropdown-item btn btn-sm rounded-0 text-secondary"
                  onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                    e.preventDefault();
                    closeDropdown();
                    props.setModalStatus({
                      open: true,
                      apiKey: props.apiKey,
                    });
                  }}
                >
                  <div className="d-flex flex-row align-items-center">
                    <FaPencilAlt className={`mr-2 ${styles.btnIcon}`} />
                    <span>Edit</span>
                  </div>
                </button>

                <button
                  data-testid="deleteAPIKeyModalBtn"
                  className="dropdown-item btn btn-sm rounded-0 text-secondary"
                  onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                    e.preventDefault();
                    closeDropdown();
                    setDeletionModalStatus(true);
                  }}
                >
                  <div className="d-flex flex-row align-items-center">
                    <FaTrashAlt className={`mr-2 ${styles.btnIcon}`} />
                    <span>Delete</span>
                  </div>
                </button>
              </div>

              <button
                className={`btn btn-light p-0 text-secondary text-center ${styles.btnDropdown}`}
                onClick={() => setDropdownMenuStatus(true)}
              >
                <BsThreeDotsVertical />
              </button>
            </div>
          </div>

          <div className="mt-2 d-flex flex-row align-items-baseline">
            <div className="text-truncate">
              <small className="text-muted text-uppercase mr-1">API-KEY-ID: </small>
              <small>{props.apiKey.apiKeyId}</small>
            </div>
            <div className={`ml-1 ${styles.copyBtn}`}>
              <div className={`position-absolute ${styles.copyBtnWrapper}`}>
                <ButtonCopyToClipboard
                  text={props.apiKey.apiKeyId!}
                  className="btn-link border-0 text-secondary font-weight-bold"
                />
              </div>
            </div>
          </div>
          <div className="text-truncate">
            <small className="text-muted text-uppercase mr-1">Created at: </small>
            <small>{moment(props.apiKey.createdAt! * 1000).format('YYYY/MM/DD HH:mm:ss (Z)')}</small>
          </div>
        </div>
      </div>
    </div>
  );
};

export default APIKeyCard;
