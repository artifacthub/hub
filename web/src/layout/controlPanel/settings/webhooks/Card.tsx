import classnames from 'classnames';
import { isUndefined } from 'lodash';
import React, { useContext, useRef, useState } from 'react';
import { BsThreeDotsVertical } from 'react-icons/bs';
import { FaPencilAlt, FaTrashAlt } from 'react-icons/fa';
import { IoMdCloseCircle } from 'react-icons/io';
import { TiWarningOutline } from 'react-icons/ti';

import { API } from '../../../../api';
import { AppCtx } from '../../../../context/AppCtx';
import useOutsideClick from '../../../../hooks/useOutsideClick';
import { ErrorKind, Webhook } from '../../../../types';
import alertDispatcher from '../../../../utils/alertDispatcher';
import ElementWithTooltip from '../../../common/ElementWithTooltip';
import Modal from '../../../common/Modal';
import styles from './Card.module.css';
import LastNotificationsModal from './LastNotificationsModal';

interface Props {
  webhook: Webhook;
  onEdition: () => void;
  onDeletion: () => void;
  onAuthError: () => void;
}

const WebhookCard = (props: Props) => {
  const { ctx } = useContext(AppCtx);
  const [isDeleting, setIsDeleting] = useState<boolean>(false);
  const [dropdownMenuStatus, setDropdownMenuStatus] = useState<boolean>(false);
  const dropdownMenu = useRef(null);
  const [deletionModalStatus, setDeletionModalStatus] = useState<boolean>(false);

  const closeDropdown = () => {
    setDropdownMenuStatus(false);
  };

  useOutsideClick([dropdownMenu], dropdownMenuStatus, closeDropdown);

  async function deleteWebhook() {
    try {
      setIsDeleting(true);
      await API.deleteWebhook(props.webhook.webhookId!, ctx.prefs.controlPanel.selectedOrg);
      setIsDeleting(false);
      props.onDeletion();
    } catch (err) {
      setIsDeleting(false);
      if (err.kind === ErrorKind.Unauthorized) {
        props.onAuthError();
      } else {
        alertDispatcher.postAlert({
          type: 'danger',
          message: 'An error occurred deleting the webhook, please try again later.',
        });
      }
    }
  }

  return (
    <div className="col-12 col-xxl-6 py-sm-3 py-2" role="listitem">
      <div className={`card cardWithHover w-100 ${styles.card}`}>
        <div className={`card-body p-0 position-relative ${styles.body}`}>
          <div className="d-flex flex-row">
            <div className={`h5 card-title mb-3 mr-3 ${styles.title}`}>
              <div className="d-flex flex-row align-items-start">
                <div>{props.webhook.name}</div>
                {props.webhook.active ? (
                  <span
                    className={`ml-3 mt-1 font-weight-bold badge badge-pill border border-success text-success text-uppercase ${styles.badge}`}
                  >
                    Active
                  </span>
                ) : (
                  <span
                    className={`ml-3 mt-1 font-weight-bold badge badge-pill border border-dark text-dark text-uppercase ${styles.badge} ${styles.inactiveBadge}`}
                  >
                    Inactive
                  </span>
                )}
              </div>
            </div>

            {deletionModalStatus && (
              <Modal
                className={`d-inline-block ${styles.modal}`}
                closeButton={
                  <>
                    <button
                      className={`btn btn-sm btn-light text-uppercase ${styles.btnLight}`}
                      onClick={() => setDeletionModalStatus(false)}
                    >
                      <div className="d-flex flex-row align-items-center">
                        <IoMdCloseCircle className="mr-2" />
                        <span>Cancel</span>
                      </div>
                    </button>

                    <button
                      data-testid="deleteWebhookBtn"
                      className="btn btn-sm btn-danger ml-3"
                      onClick={(e) => {
                        e.preventDefault();
                        deleteWebhook();
                      }}
                      disabled={isDeleting}
                    >
                      <div className="d-flex flex-row align-items-center text-uppercase">
                        {isDeleting ? (
                          <>
                            <span className="spinner-grow spinner-grow-sm" role="status" aria-hidden="true" />
                            <span className="ml-2">Deleting...</span>
                          </>
                        ) : (
                          <>
                            <FaTrashAlt className={`mr-2 ${styles.btnDeleteIcon}`} />
                            <span>Delete</span>
                          </>
                        )}
                      </div>
                    </button>
                  </>
                }
                header={<div className={`h3 m-2 flex-grow-1 ${styles.title}`}>Delete webhook</div>}
                onClose={() => setDeletionModalStatus(false)}
                open
              >
                <div className="mt-3 mw-100 text-center">
                  <p>Are you sure you want to delete this webhook?</p>
                </div>
              </Modal>
            )}

            <div className="ml-auto">
              <div
                ref={dropdownMenu}
                className={classnames('dropdown-menu dropdown-menu-right p-0', styles.dropdownMenu, {
                  show: dropdownMenuStatus,
                })}
              >
                <div className={`arrow ${styles.arrow}`} />

                <button
                  data-testid="editWebhookBtn"
                  className="dropdown-item btn btn-sm rounded-0 text-secondary"
                  onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                    e.preventDefault();
                    closeDropdown();
                    props.onEdition();
                  }}
                >
                  <div className="d-flex flex-row align-items-center">
                    <FaPencilAlt className={`mr-2 ${styles.btnIcon}`} />
                    <span>Edit</span>
                  </div>
                </button>

                <button
                  data-testid="deleteWebhookModalBtn"
                  className="dropdown-item btn btn-sm rounded-0 text-secondary"
                  onClick={(e: React.MouseEvent<HTMLButtonElement>) => {
                    e.preventDefault();
                    closeDropdown();
                    setDeletionModalStatus(true);
                  }}
                >
                  <div className="d-flex flex-row align-items-center">
                    <FaTrashAlt className={`mr-2 ${styles.btnIcon}`} />
                    <span>Delete</span>
                  </div>
                </button>
              </div>

              <button
                className={`btn btn-light p-0 text-secondary text-center ${styles.btnDropdown}`}
                onClick={() => setDropdownMenuStatus(true)}
              >
                <BsThreeDotsVertical />
              </button>
            </div>
          </div>

          <div className="d-flex flex-column">
            <div className="card-subtitle d-flex flex-column mw-100 mt-1">
              <p className="card-text">{props.webhook.description}</p>
            </div>

            <div className="text-truncate">
              <small className="text-muted text-uppercase mr-2">Url:</small>
              <small>{props.webhook.url}</small>
            </div>

            <div className="d-flex flex-row justify-content-between align-items-baseline">
              {props.webhook.lastNotifications && (
                <div className="d-none d-md-inline mt-2">
                  <LastNotificationsModal notifications={props.webhook.lastNotifications} />
                </div>
              )}

              {(isUndefined(props.webhook.packages) || props.webhook.packages.length === 0) && (
                <div className="ml-auto mt-2">
                  <ElementWithTooltip
                    element={
                      <span
                        className={`d-flex flex-row align-items-center badge badge-warning badge-pill ${styles.badgeNoPackages}`}
                      >
                        <TiWarningOutline />
                        <span className="ml-1">No packages</span>
                      </span>
                    }
                    tooltipMessage="This webhook is not associated to any packages."
                    active
                    visibleTooltip
                  />
                </div>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default WebhookCard;
