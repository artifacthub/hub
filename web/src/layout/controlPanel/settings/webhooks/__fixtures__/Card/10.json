{
  "webhookId": "d1e1299b-c3ce-4618-b67b-f5e63f4b891f",
  "name": "Webhook name",
  "description": "this is a test description",
  "url": "http://jksdhfkjfds",
  "contentType": "application/json",
  "template": "{\n    \"event\": {\n        \"id\": \"{{ .Event.ID }}\",\n        \"kind\": {{ .Event.Kind }}\n    },\n    \"package\": {\n        \"kind\": {{ .Package.Kind }},\n        \"name\": \"{{ .Package.Name }}\",\n        \"version\": \"{{ .Package.Version }}\",\n        \"publisher\": \"{{ .Package.Publisher }}\",\n        \"url\": \"{{ .Package.URL }}\"\n    }\n}",
  "active": true,
  "eventKinds": [0],
  "packages": [
    {
      "packageId": "701964f8-1730-45d2-bdfa-ae2b2d0ecac6",
      "name": "envoy",
      "logoImageId": "c5cfce82-9b0c-4a26-9a8d-c9d0fd5aa2cc",
      "repository": {
        "repositoryId": "d3945a69-2aae-4ff9-8fb3-ce3b070d0384",
        "kind": 0,
        "name": "stable",
        "displayName": "Stable",
        "organizationName": "helm",
        "organizationDisplayName": "Helm"
      }
    }
  ],
  "lastNotifications": [
    {
      "notificationId": "0f1c1a3f-55f7-4d7c-91e6-8ed081a1f400",
      "createdAt": 1590852609,
      "processed": true,
      "processedAt": 1590852610,
      "error": "template: :2:20: executing \"\" at <.kjsk>: can't evaluate field kjsk in type *hub.NotificationTemplateData"
    }
  ]
}
