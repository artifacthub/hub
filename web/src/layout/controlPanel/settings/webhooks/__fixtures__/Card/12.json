{
  "webhookId": "d1e1299b-c3ce-4618-b67b-f5e63f4b891f",
  "name": "Webhook name",
  "description": "this is a test description",
  "url": "http://jksdhfkjfds",
  "contentType": "application/json",
  "template": "{\n    \"event\": {\n        \"id\": \"{{ .Event.ID }}\",\n        \"kind\": {{ .Event.Kind }}\n    },\n    \"package\": {\n        \"kind\": {{ .Package.Kind }},\n        \"name\": \"{{ .Package.Name }}\",\n        \"version\": \"{{ .Package.Version }}\",\n        \"publisher\": \"{{ .Package.Publisher }}\",\n        \"url\": \"{{ .Package.URL }}\"\n    }\n}",
  "active": true,
  "eventKinds": [0]
}
