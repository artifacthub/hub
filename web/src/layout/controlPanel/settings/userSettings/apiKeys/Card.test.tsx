import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { mocked } from 'ts-jest/utils';

import { API } from '../../../../../api';
import { APIKey, ErrorKind } from '../../../../../types';
import alertDispatcher from '../../../../../utils/alertDispatcher';
import Card from './Card';
jest.mock('../../../../../api');
jest.mock('../../../../../utils/alertDispatcher');
jest.mock('moment', () => () => ({ format: () => '2020/06/18 16:35:39 (+00:00)' }));

const APIKeyMock: APIKey = {
  apiKeyId: 'bf28013f-610e-4691-80a2-bd3a673c4b3f',
  name: 'key1',
  createdAt: 1592498139,
};

const setModalStatusMock = jest.fn();
const onAuthErrorMock = jest.fn();

const defaultProps = {
  apiKey: APIKeyMock,
  setModalStatus: setModalStatusMock,
  onSuccess: jest.fn(),
  onAuthError: onAuthErrorMock,
};

describe('API key Card - API keys section', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const result = render(<Card {...defaultProps} />);
    expect(result.asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      const { getByText, getByTestId } = render(<Card {...defaultProps} />);

      expect(getByTestId('APIKeyCard')).toBeInTheDocument();
      expect(getByText(APIKeyMock.name!)).toBeInTheDocument();
      expect(getByText('Created at:')).toBeInTheDocument();
      expect(getByTestId('updateAPIKeyBtn')).toBeInTheDocument();
      expect(getByTestId('deleteAPIKeyModalBtn')).toBeInTheDocument();
    });

    it('calls deleteAPIKey when leave button in dropdown is clicked', async () => {
      const { getByTestId, getByText } = render(<Card {...defaultProps} />);

      const modalBtn = getByTestId('deleteAPIKeyModalBtn');
      expect(modalBtn).toBeInTheDocument();
      fireEvent.click(modalBtn);

      expect(getByText('Are you sure you want to remove this API key?')).toBeInTheDocument();

      const btn = getByTestId('deleteAPIKeyBtn');
      fireEvent.click(btn);

      await waitFor(() => {
        expect(API.deleteAPIKey).toHaveBeenCalledTimes(1);
      });
    });

    it('calls setModalStatusMock when Edit button is clicked', () => {
      const { getByTestId } = render(<Card {...defaultProps} />);

      const btn = getByTestId('updateAPIKeyBtn');
      expect(btn).toBeInTheDocument();

      fireEvent.click(btn);
      expect(setModalStatusMock).toHaveBeenCalledTimes(1);
      expect(setModalStatusMock).toHaveBeenCalledWith({
        open: true,
        apiKey: APIKeyMock,
      });
    });
  });

  describe('on deleteAPIKey error', () => {
    it('displays generic error', async () => {
      mocked(API).deleteAPIKey.mockRejectedValue({
        kind: ErrorKind.Other,
      });
      const { getByTestId } = render(<Card {...defaultProps} />);

      const modalBtn = getByTestId('deleteAPIKeyModalBtn');
      expect(modalBtn).toBeInTheDocument();
      fireEvent.click(modalBtn);

      const btn = getByTestId('deleteAPIKeyBtn');
      fireEvent.click(btn);

      await waitFor(() => {
        expect(API.deleteAPIKey).toHaveBeenCalledTimes(1);
      });

      expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
      expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
        type: 'danger',
        message: 'An error occurred deleting the API key, please try again later.',
      });
    });

    it('calls onAuthError', async () => {
      mocked(API).deleteAPIKey.mockRejectedValue({
        kind: ErrorKind.Unauthorized,
      });
      const { getByTestId } = render(<Card {...defaultProps} />);

      const modalBtn = getByTestId('deleteAPIKeyModalBtn');
      expect(modalBtn).toBeInTheDocument();
      fireEvent.click(modalBtn);

      const btn = getByTestId('deleteAPIKeyBtn');
      fireEvent.click(btn);

      await waitFor(() => {
        expect(API.deleteAPIKey).toHaveBeenCalledTimes(1);
      });

      expect(onAuthErrorMock).toHaveBeenCalledTimes(1);
    });
  });
});
