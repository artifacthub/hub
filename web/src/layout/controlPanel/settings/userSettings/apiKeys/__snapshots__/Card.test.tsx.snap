// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API key Card - API keys section creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-12 col-xxl-6 py-sm-3 py-2"
    data-testid="APIKeyCard"
  >
    <div
      class="card h-100"
    >
      <div
        class="card-body d-flex flex-column h-100"
      >
        <div
          class="d-flex flex-row w-100 justify-content-between"
        >
          <div
            class="h5 mb-1 mr-2 titleCard"
          >
            key1
          </div>
          <div
            class="ml-auto"
          >
            <div
              class="dropdown-menu dropdown-menu-right p-0 dropdownMenu"
            >
              <div
                class="arrow arrow"
              />
              <button
                class="dropdown-item btn btn-sm rounded-0 text-secondary"
                data-testid="updateAPIKeyBtn"
              >
                <div
                  class="d-flex flex-row align-items-center"
                >
                  <svg
                    class="mr-2 btnIcon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                    />
                  </svg>
                  <span>
                    Edit
                  </span>
                </div>
              </button>
              <button
                class="dropdown-item btn btn-sm rounded-0 text-secondary"
                data-testid="deleteAPIKeyModalBtn"
              >
                <div
                  class="d-flex flex-row align-items-center"
                >
                  <svg
                    class="mr-2 btnIcon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                    />
                  </svg>
                  <span>
                    Delete
                  </span>
                </div>
              </button>
            </div>
            <button
              class="btn btn-light p-0 text-secondary text-center btnDropdown"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M9.5 13a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zm0-5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zm0-5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="mt-2 d-flex flex-row align-items-baseline"
        >
          <div
            class="text-truncate"
          >
            <small
              class="text-muted text-uppercase mr-1"
            >
              API-KEY-ID: 
            </small>
            <small>
              bf28013f-610e-4691-80a2-bd3a673c4b3f
            </small>
          </div>
          <div
            class="ml-1 copyBtn"
          >
            <div
              class="position-absolute copyBtnWrapper"
            >
              <div
                class="position-relative undefined"
              >
                <button
                  class="btn btn-sm btn-link border-0 text-secondary font-weight-bold"
                  data-testid="ctcBtn"
                  type="button"
                >
                  <div
                    class="d-flex flex-row align-items-center"
                  >
                    <svg
                      fill="none"
                      height="1em"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <rect
                        height="13"
                        rx="2"
                        ry="2"
                        width="13"
                        x="9"
                        y="9"
                      />
                      <path
                        d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
                      />
                    </svg>
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="text-truncate"
        >
          <small
            class="text-muted text-uppercase mr-1"
          >
            Created at: 
          </small>
          <small>
            2020/06/18 16:35:39 (+00:00)
          </small>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
